executor/infoschema_reader.go

```diff
@@ -1869,6 +1874,82 @@ func (e *memtableRetriever) setDataForPlacementPolicy(ctx sessionctx.Context) er
 	return nil
 }
 
+func (e *memtableRetriever) setDataForClientErrorsSummary(ctx sessionctx.Context, tableName string) error {
+	// Seeing client errors should require the PROCESS privilege, with the exception of errors for your own user.
+	// This is similar to information_schema.processlist, which is the closest comparison.
+	var hasProcessPriv bool
+	loginUser := ctx.GetSessionVars().User
+	if pm := privilege.GetPrivilegeManager(ctx); pm != nil {
+		if pm.RequestVerification(ctx.GetSessionVars().ActiveRoles, "", "", "", mysql.ProcessPriv) {
+			hasProcessPriv = true
+		}
+	}
+
+	var rows [][]types.Datum
+	switch tableName {
+	case infoschema.TableClientErrorsSummaryGlobal:
+		if !hasProcessPriv {
+			return plannercore.ErrSpecificAccessDenied.GenWithStackByArgs("PROCESS")
+		}
+		for code, summary := range errno.GlobalStats() {
```

Comment:
You only lock the global stats for getting the reference but do not lock it when iterating the map. Is that right?
I assume it may lead to concurrent access to the map.

Reply:
You are right, thanks! I have changed it so that the infoschema-reader is now responsible for holding the mutex, but I am not sure if it is better to copy the map instead? Let me know what you prefer.

Reply:
I think deep copy before iterating will be better.
Besides, I think we need only one lock for all the 3 maps, instead of one lock for each map.

Reply:
I have made this change.

I have also removed the mutex protecting each individual `ErrorSummary`. I am not sure if this is what you meant, I can revert https://github.com/pingcap/tidb/pull/22382/commits/3c960e921214c871ecedc6e0d38dbe96d13ec519 if necessary. It will only be in very abnormal scenarios that this mutex will ever be hot.